pkg/tfbridge/validate_input_types.go

```diff
@@ -0,0 +1,361 @@
+			return nil
+		}
+
+		// don't validate types with Discriminator (yet)
+		if typeSpec.Discriminator != nil {
+			return nil
+		}
+
+		// default type
+		if typeSpec.Type != "" {
+			typeSpec.OneOf = append(typeSpec.OneOf, pschema.TypeSpec{Type: typeSpec.Type})
+		}
+
+		failures := []TypeFailure{}
+		for _, spec := range typeSpec.OneOf {
+			failure := v.validatePropertyValue(propertyValue, spec, propertyPath)
+			if len(failure) == 0 {
+				// we have found the correct type
+				return nil
+			}
+			failures = append(failures, failure...)
+		}
+		return failures
+	}
+
+	switch typeSpec.Type {
+	case "bool":
+		// The bridge permits coalescing strings to booleans, hence skip strings.
+		if !propertyValue.IsBool() && !propertyValue.IsString() {
+			return []TypeFailure{{
+				ResourcePath: propertyPath.String(),
+				Reason: fmt.Sprintf(
+					"expected boolean type, got %s type",
+					propertyValue.TypeString(),
+				),
+			}}
+		}
+		return nil
+	case "integer", "number":
+		// The bridge permits coalescing strings to numbers, hence skip strings.
+		if !propertyValue.IsNumber() && !propertyValue.IsString() {
+			return []TypeFailure{{
+				ResourcePath: propertyPath.String(),
+				Reason: fmt.Sprintf(
+					"expected number type, got %s type",
+					propertyValue.TypeString(),
+				),
+			}}
+		}
+		return nil
+	case "string":
+		// The bridge permits coalescing numbers and booleans to strings, hence skip these.
+		if !propertyValue.IsString() && !propertyValue.IsNumber() && !propertyValue.IsBool() {
+			return []TypeFailure{{
+				ResourcePath: propertyPath.String(),
+				Reason: fmt.Sprintf(
+					"expected string type, got %s type",
+					propertyValue.TypeString(),
+				),
+			}}
+		}
+		return nil
+	case "array":
+		if propertyValue.IsArray() {
+			if typeSpec.Items == nil {
+				// Unknown item type so nothing more to check.
+				return nil
+			}
+			// Check every item against the array element type.
+			failures := []TypeFailure{}
+			for idx, arrayValue := range propertyValue.ArrayValue() {
+				pb := append(propertyPath, idx)
+				failure := v.validatePropertyValue(arrayValue, *typeSpec.Items, pb)
+				if failure != nil {
+					failures = append(failures, failure...)
+				}
+			}
+			return failures
+		}
+		return []TypeFailure{{
+			ResourcePath: propertyPath.String(),
+			Reason: fmt.Sprintf(
+				"expected array type, got %s type",
+				propertyValue.TypeString(),
+			),
+		}}
+	case "object":
+		// This is not really an object but a map type with some element type, which is assumed to be string if
+		// unspecified. Check accordingly. This should be very similar to the "array" case.
+		//
+		// elementTypeSpec := typeSpec.AdditionalProperties
+		if propertyValue.IsObject() {
+			if typeSpec.AdditionalProperties == nil {
+				// Unknown item type so nothing more to check
+				return nil
+			}
+			// if there is a ref to another type, get that and then validate
+			if typeSpec.AdditionalProperties.Ref != "" {
```

Comment:
You don't need this if clause at all, I think. What's going to happen is that v.validatePropertyValue on *typeSpec.AdditionalProperties is going to take full care of this.

Reply:
I see it's under test, checking up on it, hm..

Reply:
Removing this block makes TestValidateInputType_arrays/object_double_nested_object_type_success4 fail, I've instrumented that test case and I think it's actually not asserting the right behavior, with this block removed I think the validator correctly fails on that test case. Step by step:

```
=== RUN   TestValidateInputType_arrays/object_double_nested_object_type_success

# validating against array of objects
validatePropertyValue {[{map[prop:{map[objectStringProp:{foo}]}]}]} object_double_nested_object_type_success4 {
  "type": "array",
  "items": {
    "$ref": "#/types/pkg:index/type:ObjectDoubleNestedObjectType"
  }
}

# validating first element against an object
validatePropertyValue {map[prop:{map[objectStringProp:{foo}]}]} object_double_nested_object_type_success4[0] {
  "$ref": "#/types/pkg:index/type:ObjectDoubleNestedObjectType"
}

# down to prop attribute of the object, validating against a map of objects
validatePropertyValue {map[objectStringProp:{foo}]} object_double_nested_object_type_success4[0].prop {
  "type": "object",
  "additionalProperties": {
    "type": "object",
    "$ref": "#/types/pkg:index/type:ObjectStringType"
  }
}

# down objectStringProp map key, validating "foo" string against an object type
validatePropertyValue {foo} object_double_nested_object_type_success4[0].prop.objectStringProp {
  "type": "object",
  "$ref": "#/types/pkg:index/type:ObjectStringType"
}

# this should fail as it does 
    validate_input_types_test.go:1552: 0 failures, got 1: [{expected object type, got string type object_double_nested_object_type_success4[0].prop.objectStringProp}]
--- FAIL: TestValidateInputType_arrays (0.00s)
    --- FAIL: TestValidateInputType_arrays/object_double_nested_object_type_success4 (0.00s)
```

Reply:
E.g. for the double nested object type, TypeScript would have to supply something like this:

```
{"prop":  {"key1": {"objectStringProp": "A"}, "key2": {"objectStringProp": "B"}}}
```

Reply:
We can always check by generating a TypeScript SDK for a schema that looks like that.

Reply:
Ok I remove that block